### PR TITLE
util/metrics: add thread context switch metrics

### PR DESCRIPTION
**Pull request by 5kbpers**
## What have you changed? (mandatory)

This PR adds new metrics to measure context switches of each thread.

## What are the type of the changes? (mandatory)

- Improvement (change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

Create a cluster by ansible and check Grafana panels.

## Does this PR affect documentation (docs) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

Yes. I have opened a [PR](https://github.com/pingcap/tidb-ansible/pull/687) to tidb-ansible(add two panels to Grafana).

## Refer to a related PR or issue link (optional)

<https://github.com/tikv/tikv/issues/4273>

### CODE REVIEW

File: src/util/metrics/threads_linux.rs [hunk 1]
```diff
@@ -143,11 +155,25 @@ impl Collector for ThreadsCollector {
                         write_total.inc_by(write_delta);
                     }
                 }
+
+                if let Ok(status) = pid::status_task(self.pid, tid) {
+                    let total = status.voluntary_ctxt_switches + status.nonvoluntary_ctxt_switches;
```
Comment (overvenus) [inline src/util/metrics/threads_linux.rs]: Could you split them into two metrics?

**Comment by AndreMouche** [comment]
Please update the description of this PR.

File: src/util/metrics/threads_linux.rs [hunk 2]
```diff
@@ -32,6 +32,8 @@ struct Metrics {
     cpu_totals: CounterVec,
     io_totals: CounterVec,
     threads_state: IntGaugeVec,
+    voluntary_ctxt_switches: CounterVec,
```
Comment (overvenus) [inline src/util/metrics/threads_linux.rs]: How about `IntCounterVec`?

File: src/util/metrics/threads_linux.rs [hunk 3]
```diff
@@ -143,11 +166,37 @@ impl Collector for ThreadsCollector {
                         write_total.inc_by(write_delta);
                     }
                 }
+
```
Comment (AndreMouche) [inline src/util/metrics/threads_linux.rs]: Could we create an independent function to collect the context switch metrics?

**Comment by siddontang** [comment]
How does the Grafana represent for this metric?

**Comment by 5kbpers** [comment]
> How does the Grafana represent for this metric?

I add two panels "Thread Voluntary Context Switch" and "Thread Nonvoluntary Context Switch" to Thread row.

File: src/util/metrics/threads_linux.rs (hunk 3 above)
Comment (5kbpers) [reply src/util/metrics/threads_linux.rs]: > Could we create an independent function to collect the context switch metrics?

I think we should create a function for each metric if we do this......

Comment (AndreMouche) [reply src/util/metrics/threads_linux.rs]: got it

**Pull request merged**